File: lib/cortexadb/examples/python/basic_usage.py
"""
CortexaDB Python Example - Basic Usage

Demonstrates core features:
- Opening a database
- Storing memories with embeddings (Unified .add)
- High-performance .ingest (batching)
- Fluent .query (QueryBuilder)
- Scoped .collection support
- Graph relationships
"""

from cortexadb import CortexaDB, HashEmbedder
import os
import shutil


def main():
    db_path = "example_agent.db"

    # Cleanup old db directory
    if os.path.isdir(db_path):
        shutil.rmtree(db_path)

    print("=== CortexaDB Python Example (v0.1.8) ===\n")

    # 1. Open database with embedder (auto-embeds text)
    # HashEmbedder generates deterministic embeddings for testing
    db = CortexaDB.open(db_path, embedder=HashEmbedder(dimension=128))
    print(f"Opened: {db}")

    # -----------------------------------------------------------
    # 2. Unified Add (stores a memory)
    # -----------------------------------------------------------
    print("\n[1] Adding information...")
    m1 = db.add(
        "The user lives in Paris and loves baguette.", metadata={"category": "personal"}
    )
    m2 = db.add("Paris is the capital of France.", metadata={"category": "fact"})
    m3 = db.add(
        "The weather in Paris is often rainy in autumn.",
        metadata={"category": "weather"},
    )
    print(f"   Stored 3 memories: IDs {m1}, {m2}, {m3}")

    # -----------------------------------------------------------
    # 3. High-Performance Ingest (100x faster batching)
    # -----------------------------------------------------------
    print("\n[2] Ingesting text with batching...")

    # Recursive (default) - splits paragraphs → sentences → words
    long_text = """
    First paragraph with some content.
    
    Second paragraph with more details.
    
    Third paragraph to complete the example.
    """
    # v0.1.8 uses optimized batch insertion internally
    ids = db.ingest(long_text, strategy="recursive", chunk_size=100, overlap=10)
    print(f"   Recursive batching: {len(ids)} chunks stored in ms")

    # -----------------------------------------------------------
    # 4. Fluent Query Builder
    # -----------------------------------------------------------
    print("\n[3] Using Fluent Query Builder...")
    
    results = db.query("Where does the user live?") \
        .limit(3) \
        .execute()
        
    print(f"   Query: 'Where does the user live?'")
    for res in results:
        print(f"   - ID: {res.id}, Score: {res.score:.4f}")

    # -----------------------------------------------------------
    # 5. Graph Relationships
    # -----------------------------------------------------------
    print("\n[4] Creating graph connections...")
    db.connect(m1, m2, "related_to")
    db.connect(m2, m3, "mentioned_in")
    print(f"   Connected memories: {m1} → {m2} → {m3}")

    # -----------------------------------------------------------
    # 6. Collections (Namespaced isolation)
    # -----------------------------------------------------------
    print("\n[5] Using Collections...")

    travel = db.collection("travel_agent")
    travel.add("Flight to Tokyo booked for June.")
    travel.add("Hotel reservation confirmed.")

    # Search scoped to the collection
    results = travel.search("Tokyo")
    print(f"   Travel collection: {len(results)} results for 'Tokyo'")

    # Or use QueryBuilder from a collection
    scoped_results = travel.query("Tokyo").limit(1).execute()
    print(f"   Scoped QueryBuilder: {len(scoped_results)} result")

    # -----------------------------------------------------------
    # 7. Stats
    # -----------------------------------------------------------
    print("\n[6] Database stats...")
    stats = db.stats()
    print(f"   Total entries: {stats.entries}")
    print(f"   Indexed embeddings: {stats.indexed_embeddings}")

    # Cleanup (database flushes on __exit__ or delete)
    del db
    if os.path.isdir(db_path):
        shutil.rmtree(db_path)

    print("\n=== Example Complete! ===")


if __name__ == "__main__":
    main()

File: lib/cortexadb/examples/rust/basic_usage.rs
//! CortexaDB Rust Example - Basic Usage
//!
//! Demonstrates core features:
//! - Opening a database
//! - Storing memories
//! - Using chunking strategies
//! - Hybrid search
//! - Graph relationships
//! - Collection-scoped operations

use cortexadb_core::{chunk, ChunkingStrategy, CortexaDB};
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::path::Path;

fn embed_text(text: &str, dimension: usize) -> Vec<f32> {
    let mut vec = vec![0.0_f32; dimension];

    for (i, token) in text.split_whitespace().enumerate() {
        let mut hasher = DefaultHasher::new();
        token.to_lowercase().hash(&mut hasher);
        let hash = hasher.finish() as usize;
        let idx = hash % dimension;
        let sign = if hash & 1 == 0 { 1.0 } else { -1.0 };
        vec[idx] += sign;
        vec[(idx + i) % dimension] += 0.25 * sign;
    }

    let norm = vec.iter().map(|v| v * v).sum::<f32>().sqrt();
    if norm > 0.0 {
        for value in &mut vec {
            *value /= norm;
        }
    }

    vec
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let db_path = "example_rust.db";
    let dimension = 64;

    // Cleanup old database directory
    if Path::new(db_path).exists() {
        std::fs::remove_dir_all(db_path)?;
    }

    println!("=== CortexaDB Rust Example ===\n");

    // -----------------------------------------------------------
    // 1. Open Database (current API)
    // -----------------------------------------------------------
    let db = CortexaDB::open(db_path, dimension)?;
    println!("Opened database: {} entries", db.stats()?.entries);

    // -----------------------------------------------------------
    // 2. Chunking Strategies (5 available)
    // -----------------------------------------------------------
    println!("\n[1] Using chunking strategies...");

    let text = r#"First paragraph with some content here.
    
Second paragraph with more details.
    
Third paragraph to complete the example."#;

    // Recursive (default for RAG)
    let chunks = chunk(text, ChunkingStrategy::Recursive { chunk_size: 50, overlap: 5 });
    println!("   Recursive: {} chunks", chunks.len());

    // Semantic (split by paragraphs)
    let chunks = chunk(text, ChunkingStrategy::Semantic { overlap: 2 });
    println!("   Semantic: {} chunks", chunks.len());

    // Fixed (character-based)
    let chunks = chunk(text, ChunkingStrategy::Fixed { chunk_size: 30, overlap: 3 });
    println!("   Fixed: {} chunks", chunks.len());

    // -----------------------------------------------------------
    // 3. Markdown Chunking
    // -----------------------------------------------------------
    println!("\n[2] Markdown chunking...");

    let md_text = r#"# Heading 1

Content under heading 1.

## Heading 2

Content under heading 2.

### Heading 3

Content under heading 3.
"#;

    let chunks = chunk(md_text, ChunkingStrategy::Markdown { preserve_headers: true, overlap: 1 });
    println!("   Markdown: {} chunks", chunks.len());
    for (i, chunk) in chunks.iter().take(2).enumerate() {
        println!("      Chunk {}: {}...", i, &chunk.text[..chunk.text.len().min(30)]);
    }

    // -----------------------------------------------------------
    // 4. JSON Chunking
    // -----------------------------------------------------------
    println!("\n[3] JSON chunking...");

    let json_text = r#"{"user": {"name": "John", "age": 30}, "city": "Paris"}"#;

    let chunks = chunk(json_text, ChunkingStrategy::Json { overlap: 0 });
    println!("   JSON: {} key-value pairs", chunks.len());
    for chunk in &chunks {
        if let Some(ref meta) = chunk.metadata {
            println!("      {} = {}", meta.key.as_ref().unwrap(), meta.value.as_ref().unwrap());
        }
    }

    // -----------------------------------------------------------
    // 5. High-Performance Batch Storage
    // -----------------------------------------------------------
    println!("\n[4] Storing memories (Batch mode)...");

    let text1 = "The user lives in Paris and loves programming.";
    let text2 = "CortexaDB is a vector database for AI agents.";
    let text3 = "Rust is a systems programming language.";

    use cortexadb_core::BatchRecord;

    let records = vec![
        BatchRecord {
            collection: "default".to_string(),
            content: text1.as_bytes().to_vec(),
            embedding: Some(embed_text(text1, dimension)),
            metadata: None,
        },
        BatchRecord {
            collection: "default".to_string(),
            content: text2.as_bytes().to_vec(),
            embedding: Some(embed_text(text2, dimension)),
            metadata: None,
        },
        BatchRecord {
            collection: "default".to_string(),
            content: text3.as_bytes().to_vec(),
            embedding: Some(embed_text(text3, dimension)),
            metadata: None,
        },
    ];

    // Bulk insert with 100x speedup
    let last_id = db.add_batch(records)?;
    println!("   Batch finished. Last inserted ID: {}", last_id.last().unwrap());
    
    // For manual IDs in the example, we'll use 1, 2, 3 assuming clean start
    let id1 = 1; let id2 = 2; let id3 = 3;

    // -----------------------------------------------------------
    // 6. Graph Relationships
    // -----------------------------------------------------------
    println!("\n[5] Creating graph connections...");

    db.connect(id1, id2, "uses")?;
    db.connect(id2, id3, "written_in")?;

    println!("   Connected: {} -> {} -> {}", id1, id2, id3);

    // -----------------------------------------------------------
    // 7. Search (query embedding -> top-k results)
    // -----------------------------------------------------------
    println!("\n[6] Querying memories...");
    let query = "Where does the user live?";
    let hits = db.search(embed_text(query, dimension), 3, None)?;
    for hit in hits {
        let mem = db.get_memory(hit.id)?;
        let content = String::from_utf8_lossy(&mem.content);
        println!("   - ID: {}, Score: {:.4}, Content: {}", hit.id, hit.score, content);
    }

    // -----------------------------------------------------------
    // 8. Collection-scoped retrieval
    // -----------------------------------------------------------
    println!("\n[7] Collections...");
    let travel_text = "Flight to Tokyo booked for June.";
    let col_id = db.add_with_content(
        "travel_agent",
        travel_text.as_bytes().to_vec(),
        embed_text(travel_text, dimension),
        None,
    )?;
    println!("   Stored in collection 'travel_agent': ID {}", col_id);
    let col_hits = db.search_in_collection(
        "travel_agent",
        embed_text("Tokyo travel plans", dimension),
        5,
        None,
    )?;
    println!("   travel_agent hits: {}", col_hits.len());

    // -----------------------------------------------------------
    // 9. Stats
    // -----------------------------------------------------------
    println!("\n[8] Database stats...");
    let stats = db.stats()?;
    println!("   Entries: {}", stats.entries);
    println!("   Indexed embeddings: {}", stats.indexed_embeddings);

    println!("\n=== Example Complete! ===");

    // Cleanup
    if Path::new(db_path).exists() {
        std::fs::remove_dir_all(db_path)?;
    }

    Ok(())
}
